### PR TITLE
agent: correct CPUShares and CPUWeight value

### PR DESCRIPTION
**Pull request by jongwu**
If cgroup driver is systemd, CPUShares, for cgroup v1, should be at least 2 [1] and CPUWeight for cgroup v2, should be at least 1 [2].

Fixes: #8340
Signed-off-by: Jianyong Wu <jianyong.wu@arm.com>

[1] https://github.com/systemd/systemd/blob/d19434fbf81db04d03c8cffa87821f754a86635b/src/basic/cgroup-util.h#L122
[2] https://github.com/systemd/systemd/blob/d19434fbf81db04d03c8cffa87821f754a86635b/src/basic/cgroup-util.h#L91

### CODE REVIEW

**Review by cmaf — APPROVED** [review]
thanks @jongwu

File: src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs
Comment (bergwolf) [inline src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs:84]: Could you add a reference to https://github.com/systemd/systemd/blob/d19434fbf81db04d03c8cffa87821f754a86635b/src/basic/cgroup-util.h#L91 like the one for CPUShares? Thanks!

Comment (jongwu) [reply src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs:84]: Thanks @bergwolf -, fixed.

**Review by bergwolf — APPROVED** [review]
lgtm, thanks!

**Review by amshinde — APPROVED** [review]

**Comment by jongwu** [comment]
Can we merge this? @bergwolf

**Comment by jongwu** [comment]
Hello, can we get this merged?

File: src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs
Comment (gkurz) [inline src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs:56]: So this is basically ignoring an invalid value received from the runtime... not correcting anything unlike the title seems to suggest. Is it appropriate to continue execution ? FWIW the memory subsystem bails out when receiving an invalid value. See https://github.com/kata-containers/kata-containers/blob/main/src/agent/rustjail/src/cgroups/systemd/subsystem/memory.rs#L42 .

**Review by gkurz — CHANGES_REQUESTED** [review]
Blocking this PR because it has a lot of approvals already but I'd really like to understand why invalid values are ignored instead of being corrected or at least logged.

@jongwu @bergwolf @cmaf @amshinde ?

**Comment by jongwu** [comment]
Thanks @gkurz -,  Updated.

File: src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs
Comment (gkurz) [reply src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs]: On second thought, cgroupsv2 treats 0 as a hint to use the default value. Maybe better to do the same in cgroupsv1 for consistency ? WDYT @jongwu @cmaf ?

Comment (jongwu) [reply src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs]: What's the default value of cpuShares? 100?

Comment (gkurz) [reply src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs]: > What's the default value of cpuShares? 100?

See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/sched/sched.h#n409

```
/*
 * A weight of 0 or 1 can cause arithmetics problems.
 * A weight of a cfs_rq is the sum of weights of which entities
 * are queued on this cfs_rq, so a weight of a entity should not be
 * too large, so as the shares value of a task group.
 * (The default weight is 1024 - so there's no practical
 *  limitation from this.)
 */
#define MIN_SHARES		(1UL <<  1)
#define MAX_SHARES		(1UL << 18)
```
You can :
- convert runtime input of 0 to 1024
- bail out if value is not in the MIN_SHARES...MAX_SHARES range


 and the value should be in the 2..262144 range.

File: src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs
Comment (gkurz) [reply src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs]: Similarly to cpu shares, cpu weight has a validity range.

See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/admin-guide/cgroup-v2.rst#n646
```
All weights are in the range [1, 10000] with the default at 100.  This
allows symmetric multiplicative biases in both directions at fine
enough granularity while staying in the intuitive range.
```

Maybe bail out if value is out of range, which happens if passed cpu shares is > 262144, for consistency.

Comment (jongwu) [reply src/agent/rustjail/src/cgroups/systemd/subsystem/cpu.rs]: Great! Changed.

**Review by gkurz — APPROVED** [review]
LGTM. Thanks @jongwu !

**Pull request merged**
